File: basic/studentLists_vec_sharedPtr/Students.cpp
#include <iostream>
#include <vector>
#include <memory>
#include <string>

using namespace std;

class Student {
public:
    Student(int id, const string& name) : id(id), name(name) {
        cout << "Student(ID: " << id << ", Name: " << name << ") constructed\n";
    }
    ~Student() {
        cout << "Student(ID: " << id << ", Name: " << name << ") destroyed\n";
    }
    int getId() const { return id; }
    string getName() const { return name; }
private:
    int id;
    string name;
};

int main() {
    vector<shared_ptr<Student>> students;
    
    // 向容器中添加学生对象
    students.push_back(make_shared<Student>(101, "Alice"));
    students.push_back(make_shared<Student>(102, "Bob"));
    students.push_back(make_shared<Student>(103, "Charlie"));

    // 输出所有学生的ID和姓名
    cout << "学生列表:\n";
    for (const auto& student : students) {
        cout << "ID: " << student->getId() << ", Name: " << student->getName() << "\n";
    }

    return 0;
}

File: basic/studentLists_vec_sharedPtr/Students.java
import java.util.ArrayList;
import java.util.List;

class Student {
    private int id;
    private String name;

    public Student(int id, String name) {
        this.id = id;
        this.name = name;
        System.out.println("Student(ID: " + id + ", Name: " + name + ") constructed");
    }

    @Override
    protected void finalize() throws Throwable {
        System.out.println("Student(ID: " + id + ", Name: " + name + ") destroyed");
        super.finalize();
    }

    public int getId() {
        return id;
    }

    public String getName() {
        return name;
    }
}

public class Main {
    public static void main(String[] args) {
        List<Student> students = new ArrayList<>();

        // 向列表中添加学生对象
        students.add(new Student(101, "Alice"));
        students.add(new Student(102, "Bob"));
        students.add(new Student(103, "Charlie"));

        // 输出所有学生的ID和姓名
        System.out.println("学生列表:");
        for (Student student : students) {
            System.out.println("ID: " + student.getId() + ", Name: " + student.getName());
        }
    }
}
